File: app/src/main/java/jp/hazuki/yuzubrowser/download/DownloadService.java
package jp.hazuki.yuzubrowser.download;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.PowerManager;
import android.os.PowerManager.WakeLock;
import android.os.Process;
import android.os.RemoteException;
import android.support.v4.app.NotificationCompat;
import android.text.TextUtils;
import android.webkit.CookieManager;
import android.widget.Toast;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.utils.ErrorReport;
import jp.hazuki.yuzubrowser.utils.FileUtils;
import jp.hazuki.yuzubrowser.utils.Logger;
import jp.hazuki.yuzubrowser.utils.PackageUtils;
import jp.hazuki.yuzubrowser.utils.net.HttpClientBuilder;
import jp.hazuki.yuzubrowser.utils.net.HttpResponseData;

public class DownloadService extends Service {
    public static final String TAG = "DownloadService";
    public static final String EXTRA_DOWNLOAD_INFO = "jp.hazuki.yuzubrowser.download.DownloadService.extra.EXTRA_DOWNLOAD_INFO";
    public static final int NOTIFICATION_INTERVAL = 1000;

    private static final List<Messenger> mObservers = new ArrayList<>();
    private Messenger mMessenger;
    private Handler mHandler;
    private DownloadInfoDatabase mDb;
    private NotificationManager mNotificationManager;
    private PowerManager mPowerManager;
    private static final List<DownloadThread> mThreadList = new LinkedList<>();

    public static final int REGISTER_OBSERVER = 0;
    public static final int UNREGISTER_OBSERVER = 1;
    public static final int UPDATE_PROGRESS = 2;
    public static final int UPDATE_STATE = 3;
    public static final int GET_DOWNLOAD_INFO = 4;
    public static final int CANCEL_DOWNLOAD = 5;

    private static class ServiceHandler extends Handler {
        @Override
        public void handleMessage(Message msg) {
            SWITCH:
            switch (msg.what) {
                case REGISTER_OBSERVER:
                    mObservers.add(msg.replyTo);
                    break;
                case UNREGISTER_OBSERVER:
                    mObservers.remove(msg.replyTo);
                    break;
                case UPDATE_PROGRESS:
                case UPDATE_STATE: {
                    Iterator<Messenger> it = mObservers.iterator();
                    while (it.hasNext()) {
                        Messenger messenger = it.next();
                        try {
                            messenger.send(Message.obtain(msg));
                        } catch (RemoteException e) {
                            ErrorReport.printAndWriteLog(e);
                            it.remove();
                        }
                    }
                }
                break;
                case GET_DOWNLOAD_INFO: {
                    List<DownloadRequestInfo> list = new ArrayList<>();
                    synchronized (mThreadList) {
                        for (DownloadThread thread : mThreadList) {
                            list.add(thread.getDownloadRequestInfo());
                        }
                    }

                    Iterator<Messenger> it = mObservers.iterator();
                    while (it.hasNext()) {
                        Messenger messenger = it.next();
                        try {
                            messenger.send(Message.obtain(null, GET_DOWNLOAD_INFO, list));
                        } catch (RemoteException e) {
                            ErrorReport.printAndWriteLog(e);
                            it.remove();
                        }
                    }
                }
                break;
                case CANCEL_DOWNLOAD: {
                    long id = (Long) msg.obj;
                    synchronized (mThreadList) {
                        for (DownloadThread thread : mThreadList) {
                            if (id == thread.getDownloadRequestInfo().getId()) {
                                thread.abort();
                                break SWITCH;
                            }
                        }
                        Logger.e(TAG, "id not found:" + id);
                    }
                }
                break;
                default:
                    super.handleMessage(msg);
                    break;
            }
        }
    }

    public static Messenger registerObserver(IBinder binder, Messenger activityMessenger) {
        Messenger messenger = new Messenger(binder);
        Message msg = Message.obtain(null, REGISTER_OBSERVER);
        msg.replyTo = activityMessenger;
        try {
            messenger.send(msg);
        } catch (RemoteException e) {
            ErrorReport.printAndWriteLog(e);
        }
        return messenger;
    }

    public static void unregisterObserver(Messenger serviceMessenger, Messenger activityMessenger) {
        Message msg = Message.obtain(null, UNREGISTER_OBSERVER);
        msg.replyTo = activityMessenger;
        try {
            serviceMessenger.send(msg);
        } catch (RemoteException e) {
            ErrorReport.printAndWriteLog(e);
        }
    }

    public static void getDownloadInfo(Messenger serviceMessenger, Messenger activityMessenger) {
        Message msg = Message.obtain(null, GET_DOWNLOAD_INFO);
        msg.replyTo = activityMessenger;
        try {
            serviceMessenger.send(msg);
        } catch (RemoteException e) {
            ErrorReport.printAndWriteLog(e);
        }
    }

    public static void cancelDownload(Messenger serviceMessenger, Messenger activityMessenger, long id) {
        Message msg = Message.obtain(null, CANCEL_DOWNLOAD);
        msg.replyTo = activityMessenger;
        msg.obj = id;
        try {
            serviceMessenger.send(msg);
        } catch (RemoteException e) {
            ErrorReport.printAndWriteLog(e);
        }
    }

    @Override
    public void onCreate() {
        super.onCreate();
        mMessenger = new Messenger(new ServiceHandler());
        mHandler = new Handler();
        mDb = new DownloadInfoDatabase(getApplicationContext());
        mNotificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        mPowerManager = (PowerManager) getSystemService(POWER_SERVICE);

        mDb.fixData();
        startForeground(1, getNotify());
    }

    private Notification getNotify() {
        Notification.Builder builder = new Notification.Builder(this);
        builder.setContentTitle(getString(R.string.download_service));
        builder.setSmallIcon(R.drawable.ic_yuzubrowser_white);
        builder.setPriority(Notification.PRIORITY_MIN);
        return builder.build();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        synchronized (mThreadList) {
            for (DownloadThread thread : mThreadList) {
                thread.forceAbort();
            }
        }
        stopForeground(true);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            DownloadRequestInfo info = (DownloadRequestInfo) intent.getSerializableExtra(EXTRA_DOWNLOAD_INFO);
            if (info != null) {
                new DownloadThread(info).start();
            }
        }
        return START_NOT_STICKY;
    }

    @Override
    public IBinder onBind(Intent arg0) {
        return mMessenger.getBinder();
    }

    public static void startDownloadService(Context context, DownloadInfo info) {
        Intent dintent = new Intent(context, DownloadService.class);
        dintent.putExtra(DownloadService.EXTRA_DOWNLOAD_INFO, info);
        context.startService(dintent);
    }

    public class DownloadThread extends Thread {
        private static final int DOWNLOAD_BUFFER_SIZE = 1024 * 10;
        private final DownloadRequestInfo mData;
        private boolean mAbort = false;
        private boolean mForceAbort = false;

        public DownloadThread(DownloadRequestInfo data) {
            mData = data;
        }

        public void forceAbort() {
            mForceAbort = true;
            mAbort = true;
        }

        public void abort() {
            mForceAbort = false;
            mAbort = true;
        }

        public DownloadRequestInfo getDownloadRequestInfo() {
            return mData;
        }

        @Override
        public void run() {
            Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND);

            HttpClientBuilder httpClient = HttpClientBuilder.createInstance(mData.getUrl());
            if (httpClient == null) {
                showToast("HttpClientBuilder is null");
                return;
            }

            OutputStream outputStream = null;
            InputStream inputStream = null;
            NotificationCompat.Builder notification = new NotificationCompat.Builder(getApplicationContext());
            WakeLock wakelock = null;

            long id = mDb.insert(mData);
            if (id < 0) {
                showToast("DownloadInfoDatabase#insert failed");
                return;
            }

            try {
                synchronized (mThreadList) {
                    mThreadList.add(this);
                }

                wakelock = mPowerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "DownloadThread");
                wakelock.acquire();

                notification.setSmallIcon(android.R.drawable.stat_sys_download);
                notification.setOngoing(true);
                notification.setContentTitle(mData.getFile().getName());
                //notification.setContentText(mData.getUrl());
                notification.setWhen(mData.start_time);
                notification.setProgress(0, 0, true);
                notification.setContentIntent(PendingIntent.getActivity(getApplicationContext(), 0, new Intent(getApplicationContext(), DownloadListActivity.class), 0));
                mNotificationManager.notify((int) id, notification.build());//long to int

                String cookie = CookieManager.getInstance().getCookie(mData.getUrl());
                if (!TextUtils.isEmpty(cookie)) {
                    httpClient.setHeader("Cookie", cookie);
                }

                String referer = mData.getReferer();
                if (!TextUtils.isEmpty(referer)) {
                    httpClient.setHeader("Referer", referer);
                }

                HttpResponseData response = httpClient.connect();
                if (response == null) return;

                File file = mData.getFile();
                if (file.getParentFile() != null) {
                    file.getParentFile().mkdirs();
                }

                outputStream = new FileOutputStream(file);
                inputStream = response.getInputStream();

                int max = (int) response.getContentLength();
                mData.setMaxLength(max);
                int progress = 0;
                int n;
                byte[] buffer = new byte[DOWNLOAD_BUFFER_SIZE];
                long old_sec = System.currentTimeMillis();
                long new_sec = old_sec;

                while ((n = inputStream.read(buffer)) >= 0) {
                    if (mAbort) break;

                    outputStream.write(buffer, 0, n);
                    progress += n;

                    if (new_sec > old_sec + NOTIFICATION_INTERVAL) {
                        notification.setProgress(max, progress, (max <= 0));
                        notification.setContentText(mData.getNotificationString(getApplicationContext()));
                        mNotificationManager.notify((int) id, notification.build());//long to int
                        mData.setCurrentLength(progress);

                        try {
                            mMessenger.send(Message.obtain(null, UPDATE_PROGRESS, mData));
                        } catch (RemoteException e) {
                            ErrorReport.printAndWriteLog(e);
                        }

                        old_sec = new_sec;
                    }
                    new_sec = System.currentTimeMillis();//busy?
                }

                outputStream.flush();
                outputStream.close();
                outputStream = null;
                inputStream.close();
                inputStream = null;

                if (!mAbort)
                    mData.setState(DownloadInfo.STATE_DOWNLOADED);
                else if (mForceAbort)
                    mData.setState(DownloadInfo.STATE_UNKNOWN_ERROR);
                else
                    mData.setState(DownloadInfo.STATE_CANCELED);

                mDb.updateState(mData);

                try {
                    mMessenger.send(Message.obtain(null, UPDATE_STATE, mData));
                } catch (RemoteException e) {
                    ErrorReport.printAndWriteLog(e);
                }
            } catch (IOException e) {
                ErrorReport.printAndWriteLog(e);
            } finally {
                switch (mData.getState()) {
                    case DownloadInfo.STATE_CANCELED:
                        mNotificationManager.cancel((int) id);
                        break;
                    case DownloadInfo.STATE_DOWNLOADED: {
                        notification.setOngoing(false);
                        notification.setContentTitle(mData.getFile().getName());
                        notification.setWhen(System.currentTimeMillis());
                        notification.setProgress(0, 0, false);
                        notification.setAutoCancel(true);
                        notification.setContentText(getText(R.string.download_success));
                        notification.setSmallIcon(android.R.drawable.stat_sys_download_done);

                        notification.setContentIntent(PendingIntent.getActivity(getApplicationContext(), 0, PackageUtils.createFileOpenIntent(DownloadService.this, mData.getFile()), 0));
                        mNotificationManager.notify((int) id, notification.build());

                        FileUtils.notifyImageFile(getApplicationContext(), mData.file.getAbsolutePath());
                        break;
                    }
                    case DownloadInfo.STATE_DOWNLOADING:
                        mData.setState(DownloadInfo.STATE_UNKNOWN_ERROR);
                        mDb.insert(mData);
                    /* no break */
                    default:
                        notification.setOngoing(false);
                        notification.setContentTitle(mData.getFile().getName());
                        notification.setWhen(System.currentTimeMillis());
                        notification.setProgress(0, 0, false);
                        notification.setAutoCancel(true);
                        notification.setContentText(getText(R.string.download_fail));
                        notification.setSmallIcon(android.R.drawable.stat_sys_warning);
                        notification.setContentIntent(PendingIntent.getActivity(getApplicationContext(), 0, new Intent(getApplicationContext(), DownloadListActivity.class), 0));
                        mNotificationManager.notify((int) id, notification.build());
                        break;
                }

                if (inputStream != null) {
                    try {
                        inputStream.close();
                    } catch (IOException e) {
                        ErrorReport.printAndWriteLog(e);
                    }
                }
                if (outputStream != null) {
                    try {
                        outputStream.close();
                    } catch (IOException e) {
                        ErrorReport.printAndWriteLog(e);
                    }
                }
                httpClient.destroy();

                if (wakelock != null) {
                    wakelock.release();
                }

                synchronized (mThreadList) {
                    mThreadList.remove(this);
                    if (mThreadList.isEmpty()) {
                        stopSelf();
                    }
                }
            }
        }

        private void showToast(final String str) {
            mHandler.post(new Runnable() {
                @Override
                public void run() {
                    Toast.makeText(getApplicationContext(), str, Toast.LENGTH_LONG).show();
                }
            });
        }
    }

}

File: app/src/main/java/jp/hazuki/yuzubrowser/download/service/DownloadService.kt
/*
 * Copyright (C) 2017-2018 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.download.service

import android.annotation.SuppressLint
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.Uri
import android.os.*
import android.support.annotation.StringRes
import android.support.v4.app.NotificationCompat
import android.support.v4.provider.DocumentFile
import jp.hazuki.yuzubrowser.BrowserApplication
import jp.hazuki.yuzubrowser.BuildConfig
import jp.hazuki.yuzubrowser.Constants
import jp.hazuki.yuzubrowser.R
import jp.hazuki.yuzubrowser.download.core.data.DownloadFileInfo
import jp.hazuki.yuzubrowser.download.core.data.DownloadRequest
import jp.hazuki.yuzubrowser.download.core.data.MetaData
import jp.hazuki.yuzubrowser.download.core.downloader.Downloader
import jp.hazuki.yuzubrowser.download.core.utils.createFileOpenIntent
import jp.hazuki.yuzubrowser.download.core.utils.getNotificationString
import jp.hazuki.yuzubrowser.download.core.utils.registerMediaScanner
import jp.hazuki.yuzubrowser.download.core.utils.toDocumentFile
import jp.hazuki.yuzubrowser.download.service.connection.ServiceClient
import jp.hazuki.yuzubrowser.download.service.connection.ServiceCommand
import jp.hazuki.yuzubrowser.download.service.connection.ServiceSocket
import jp.hazuki.yuzubrowser.download.ui.DownloadListActivity
import jp.hazuki.yuzubrowser.utils.ErrorReport
import jp.hazuki.yuzubrowser.utils.Logger
import jp.hazuki.yuzubrowser.utils.getPathFromUri
import org.jetbrains.anko.intentFor
import org.jetbrains.anko.longToast
import org.jetbrains.anko.toast

class DownloadService : Service(), ServiceClient.ServiceClientListener {

    private lateinit var handler: Handler
    private lateinit var powerManager: PowerManager
    private lateinit var notificationManager: NotificationManager
    private lateinit var database: DownloadDatabase
    private lateinit var messenger: Messenger

    private val threadList = mutableListOf<DownloadThread>()
    private val observers = mutableListOf<Messenger>()

    override fun onCreate() {
        handler = Handler()
        powerManager = getSystemService(Context.POWER_SERVICE) as PowerManager
        notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        database = DownloadDatabase.getInstance(this)
        messenger = Messenger(ServiceClient(this))

        val notify = NotificationCompat.Builder(this, Constants.notification.CHANNEL_DOWNLOAD_SERVICE)
                .setContentTitle(getText(R.string.download_service))
                .setSmallIcon(R.drawable.ic_yuzubrowser_white)
                .setPriority(NotificationCompat.PRIORITY_MIN)
                .build()

        database.cleanUp()

        val filter = IntentFilter(ACTION_CANCEL_DOWNLOAD)
        filter.addAction(ACTION_PAUSE_DOWNLOAD)

        registerReceiver(notificationControl, filter, BrowserApplication.PERMISSION_MYAPP_SIGNATURE, null)

        startForeground(Int.MIN_VALUE, notify)
    }

    override fun onDestroy() {
        synchronized(threadList) {
            threadList.forEach {
                it.abort()
            }
        }
        unregisterReceiver(notificationControl)
        stopForeground(true)
    }

    override fun onBind(intent: Intent?): IBinder = messenger.binder

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent != null) {
            var thread: DownloadThread? = null
            when (intent.action) {
                ACTION_START_DOWNLOAD -> {
                    val root = intent.getParcelableExtra<Uri>(EXTRA_ROOT_URI).toDocumentFile(this)
                    val file = intent.getParcelableExtra<DownloadFile>(EXTRA_REQUEST_DOWNLOAD)
                    val metadata = intent.getParcelableExtra<MetaData?>(EXTRA_METADATA)
                    if (file != null) {
                        thread = FirstDownloadThread(root, file, metadata)
                    }
                }
                ACTION_RESTART_DOWNLOAD -> {
                    val id = intent.getLongExtra(EXTRA_ID, -1)
                    val info = database[id]
                    if (info != null) {
                        thread = ReDownloadThread(info, DownloadRequest(null, null, null))
                    }
                }
            }

            if (thread != null) {
                synchronized(threadList) {
                    threadList.add(thread)
                }
                thread.start()
            }
        }
        return START_NOT_STICKY
    }

    override fun registerObserver(messenger: Messenger) {
        observers.add(messenger)
    }

    override fun unregisterObserver(messenger: Messenger) {
        observers.remove(messenger)
    }

    override fun update(msg: Message) {
        val it = observers.iterator()
        while (it.hasNext()) {
            try {
                it.next().send(Message.obtain(msg))
            } catch (e: RemoteException) {
                ErrorReport.printAndWriteLog(e)
                it.remove()
            }
        }
    }

    override fun getDownloadInfo(messenger: Messenger) {
        val list = synchronized(threadList) { threadList.map { it.info } }

        val it = observers.iterator()
        while (it.hasNext()) {
            try {
                it.next().send(Message.obtain(null, ServiceSocket.GET_DOWNLOAD_INFO, list))
            } catch (e: RemoteException) {
                ErrorReport.printAndWriteLog(e)
                it.remove()
            }
        }
    }

    override fun cancelDownload(id: Long) {
        synchronized(threadList) {
            threadList.firstOrNull { id == it.info.id }?.cancel()
        }
    }

    override fun pauseDownload(id: Long) {
        synchronized(threadList) {
            threadList.firstOrNull { id == it.info.id }?.pause()
        }
    }

    private val notificationControl = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent?) {
            if (intent == null) return

            val id = intent.getLongExtra(EXTRA_ID, -1)
            if (id >= 0) {
                when (intent.action) {
                    ACTION_CANCEL_DOWNLOAD -> cancelDownload(id)
                    ACTION_PAUSE_DOWNLOAD -> pauseDownload(id)
                }
            }
        }
    }

    private inner class FirstDownloadThread(private val root: DocumentFile, private val file: DownloadFile, private val metadata: MetaData?) : DownloadThread(file.request) {
        override val info: DownloadFileInfo by lazy {
            DownloadFileInfo(root, file, metadata ?: MetaData(this@DownloadService, root, file.url, file.request))
        }
    }

    private inner class ReDownloadThread(override val info: DownloadFileInfo, request: DownloadRequest) : DownloadThread(request)

    private abstract inner class DownloadThread(private val request: DownloadRequest) : Thread(), Downloader.DownloadListener {
        abstract val info: DownloadFileInfo

        private val notification = NotificationCompat.Builder(this@DownloadService, Constants.notification.CHANNEL_DOWNLOAD_NOTIFY)
        private val bigTextStyle = NotificationCompat.BigTextStyle()

        private lateinit var downloader: Downloader

        private var isActionAdded = false

        @SuppressLint("WakelockTimeout")
        override fun run() {
            val wakelock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "DownloadThread")
            prepareThread(wakelock)

            if (info.id < 0) {
                database.insert(info)
            } else {
                info.state = DownloadFileInfo.STATE_DOWNLOADING
                database.update(info)
            }

            if (!info.root.exists()) {
                failedCheckFolder(info, R.string.download_failed_root_not_exists)
                endThreaded(wakelock)
                return
            } else if (!info.root.canWrite()) {
                failedCheckFolder(info, R.string.download_failed_root_not_writable)
                endThreaded(wakelock)
                return
            }

            downloader = Downloader.getDownloader(this@DownloadService, info, request)
            downloader.downloadListener = this

            downloader.download()

            endThreaded(wakelock)
        }

        fun cancel() = downloader.cancel()

        fun pause() = downloader.pause()

        fun abort() = downloader.abort()

        override fun onStartDownload(info: DownloadFileInfo) {
            database.update(info)
            notification.run {
                setSmallIcon(android.R.drawable.stat_sys_download)
                setOngoing(false)
                setContentTitle(info.name)
                setWhen(info.startTime)
                setProgress(0, 0, true)
                setContentIntent(PendingIntent.getActivity(applicationContext, 0, intentFor<DownloadListActivity>(), 0))
                notificationManager.notify(info.id.toInt(), build())
            }
            updateInfo(ServiceSocket.UPDATE, info)
        }

        override fun onFileDownloaded(info: DownloadFileInfo) {
            database.update(info)
            NotificationCompat.Builder(this@DownloadService, Constants.notification.CHANNEL_DOWNLOAD_NOTIFY).run {
                setOngoing(false)
                setContentTitle(info.name)
                setWhen(System.currentTimeMillis())
                setProgress(0, 0, false)
                setAutoCancel(true)
                setContentText(getText(R.string.download_success))
                setSmallIcon(android.R.drawable.stat_sys_download_done)
                setContentIntent(PendingIntent.getActivity(applicationContext, 0, info.createFileOpenIntent(this@DownloadService), 0))
                notificationManager.notify(info.id.toInt(), build())
            }
            updateInfo(ServiceSocket.UPDATE, info)
            info.root.findFile(info.name)
                    ?.let { getPathFromUri(it.uri) }
                    ?.let { registerMediaScanner(it) }
        }

        override fun onFileDownloadFailed(info: DownloadFileInfo, cause: String?) {
            database.update(info)
            if (cause != null) {
                handler.post { longToast(cause) }
                Logger.d("download error", cause)
            }
            NotificationCompat.Builder(this@DownloadService, Constants.notification.CHANNEL_DOWNLOAD_NOTIFY).run {
                setOngoing(false)
                setContentTitle(info.name)
                setWhen(System.currentTimeMillis())
                setProgress(0, 0, false)
                setAutoCancel(true)
                setContentText(getText(R.string.download_fail))
                setSmallIcon(android.R.drawable.stat_sys_warning)
                setContentIntent(PendingIntent.getActivity(applicationContext, 0, intentFor<DownloadListActivity>(), 0))
                notificationManager.notify(info.id.toInt(), build())
            }
            updateInfo(ServiceSocket.UPDATE, info)
        }

        override fun onFileDownloadAbort(info: DownloadFileInfo) {
            database.update(info)
            if (info.state == DownloadFileInfo.STATE_PAUSED) {
                NotificationCompat.Builder(this@DownloadService, Constants.notification.CHANNEL_DOWNLOAD_NOTIFY).run {
                    setOngoing(false)
                    setContentTitle(info.name)
                    setWhen(System.currentTimeMillis())
                    setAutoCancel(true)
                    setContentText(getText(R.string.download_paused))
                    setSmallIcon(R.drawable.ic_pause_white_24dp)
                    setContentIntent(PendingIntent.getActivity(applicationContext, 0, intentFor<DownloadListActivity>(), 0))

                    val resume = Intent(ACTION_RESTART_DOWNLOAD).apply { putExtra(EXTRA_ID, info.id) }
                    addAction(R.drawable.ic_start_white_24dp, getText(R.string.resume_download),
                            PendingIntent.getService(this@DownloadService, 0, resume, 0))
                    notificationManager.notify(info.id.toInt(), build())
                }
            } else {
                notificationManager.cancel(info.id.toInt())
            }
            updateInfo(ServiceSocket.UPDATE, info)
        }

        override fun onFileDownloading(info: DownloadFileInfo, progress: Long) {
            notification.run {
                setAction(info)
                setProgress(1000, (progress * 1000 / info.size).toInt(), info.size <= 0)
                setStyle(bigTextStyle.bigText(info.getNotificationString(applicationContext)))
                notificationManager.notify(info.id.toInt(), build())
            }
            updateInfo(ServiceSocket.UPDATE, info)
        }

        private fun setAction(info: DownloadFileInfo) {
            if (!isActionAdded) {
                isActionAdded = true
                notification.run {
                    if (info.resumable) {
                        val pause = Intent(ACTION_PAUSE_DOWNLOAD).apply { putExtra(EXTRA_ID, info.id) }
                        addAction(R.drawable.ic_pause_white_24dp, getText(R.string.pause_download),
                                PendingIntent.getBroadcast(this@DownloadService, 0, pause, PendingIntent.FLAG_UPDATE_CURRENT))
                    }

                    val cancel = Intent(ACTION_CANCEL_DOWNLOAD).apply { putExtra(EXTRA_ID, info.id) }
                    addAction(R.drawable.ic_cancel_white_24dp, getText(android.R.string.cancel),
                            PendingIntent.getBroadcast(this@DownloadService, 0, cancel, PendingIntent.FLAG_UPDATE_CURRENT))
                }
            }
        }

        private fun updateInfo(@ServiceCommand command: Int, info: DownloadFileInfo) {
            try {
                messenger.send(Message.obtain(null, command, info))
            } catch (e: RemoteException) {
                ErrorReport.printAndWriteLog(e)
            }
        }

        @SuppressLint("WakelockTimeout")
        private fun prepareThread(wakeLock: PowerManager.WakeLock) {
            wakeLock.acquire()
        }

        private fun endThreaded(wakeLock: PowerManager.WakeLock) {
            wakeLock.release()

            synchronized(threadList) {
                threadList.remove(this)
                if (threadList.isEmpty()) {
                    stopSelf()
                }
            }
        }

        private fun failedCheckFolder(info: DownloadFileInfo, @StringRes message: Int) {
            info.state = DownloadFileInfo.STATE_UNKNOWN_ERROR
            database.updateWithEmptyRoot(info)
            handler.post { toast(message) }
            NotificationCompat.Builder(this@DownloadService, Constants.notification.CHANNEL_DOWNLOAD_NOTIFY).run {
                setOngoing(false)
                setContentTitle(info.name)
                setWhen(System.currentTimeMillis())
                setProgress(0, 0, false)
                setAutoCancel(true)
                setContentText(getText(message))
                setSmallIcon(android.R.drawable.stat_sys_warning)
                setContentIntent(PendingIntent.getActivity(applicationContext, 0, intentFor<DownloadListActivity>(), 0))
                notificationManager.notify(info.id.toInt(), build())
            }
            updateInfo(ServiceSocket.UPDATE, info)
        }
    }

    companion object {
        const val ACTION_START_DOWNLOAD = "${BuildConfig.APPLICATION_ID}.action.download.start"
        const val ACTION_RESTART_DOWNLOAD = "${BuildConfig.APPLICATION_ID}.action.download.restart"
        const val EXTRA_ROOT_URI = "${BuildConfig.APPLICATION_ID}.extra.root"
        const val EXTRA_REQUEST_DOWNLOAD = "${BuildConfig.APPLICATION_ID}.extra.request.download"
        const val EXTRA_METADATA = "${BuildConfig.APPLICATION_ID}.extra.metadata"
        const val EXTRA_ID = "${BuildConfig.APPLICATION_ID}.extra.id"

        private const val ACTION_CANCEL_DOWNLOAD = "${BuildConfig.APPLICATION_ID}.action.cancel.download"
        private const val ACTION_PAUSE_DOWNLOAD = "${BuildConfig.APPLICATION_ID}.action.pause.download"
    }
}